File: 03_theil_decomposition.py
"""
03 — Theil decomposition of the national supplier market
========================================================
Project: Who Sells to the State (2026_2). Adapts 2026_12_SER conventions.

Cut (a): Theil T of adjudicated ARS amounts across suppliers, per year,
         decomposed between/within province of supplier domicile.
         Also on award counts (inflation-free by construction).
Cut (b): Shannon H of rubro composition within each province x era
         (supplier-monoculture test), normalised by ln(K).
Cut (c): Nested Theil by era: between-region (NEA vs resto) ->
         between-province within region -> within-province.

Input:  data/processed/adjudicaciones_clean.parquet
Output: tables/tab_theil_provincia_anual.csv
        tables/tab_shannon_rubros_provincia_era.csv
        tables/tab_theil_anidado_era.csv
        figures/fig_theil_temporal.png
        figures/fig_shannon_nea.png

Usage:  python 03_theil_decomposition.py
"""
from pathlib import Path

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

PROJECT = Path(__file__).parent
PROC = PROJECT / "data" / "processed"
TAB = PROJECT / "tables"
FIG = PROJECT / "figures"
TAB.mkdir(exist_ok=True)
FIG.mkdir(exist_ok=True)

NEA = {"Misiones", "Corrientes", "Chaco", "Formosa"}
ERAS = ["macri", "fernandez", "milei"]


def theil_t(x):
    x = np.asarray(x, dtype=float)
    x = x[x > 0]
    if len(x) < 2:
        return np.nan
    mu = x.mean()
    return float(np.mean((x / mu) * np.log(x / mu)))


def theil_decompose(df, value, group):
    """Return total, between, within Theil T of `value` across rows grouped by `group`."""
    d = df[df[value] > 0]
    if len(d) < 2:
        return np.nan, np.nan, np.nan
    mu = d[value].mean()
    n = len(d)
    total = theil_t(d[value])
    between = within = 0.0
    for _, g in d.groupby(group):
        ng, mug = len(g), g[value].mean()
        between += (ng / n) * (mug / mu) * np.log(mug / mu)
        tg = theil_t(g[value])
        if not np.isnan(tg):
            within += (ng / n) * (mug / mu) * tg
    return total, between, within


adj = pd.read_parquet(PROC / "adjudicaciones_clean.parquet")
adj = adj[adj["es_nuevo"] & adj["provincia"].notna()]
ars = adj[adj["moneda"] == "ARS"]

# ── Cut (a): yearly Theil across suppliers, between/within province ─────────
rows = []
for year, g in ars.groupby("ejercicio"):
    firm = (g.groupby(["cuit", "provincia"], as_index=False)
            .agg(monto=("monto", "sum"), n=("monto", "size")))
    for value in ["monto", "n"]:
        tot, bet, wit = theil_decompose(firm, value, "provincia")
        rows.append({"ejercicio": year, "variable": value, "theil_total": tot,
                     "theil_between": bet, "theil_within": wit,
                     "share_between": bet / tot if tot else np.nan,
                     "n_firmas": len(firm)})
ta = pd.DataFrame(rows)
ta.to_csv(TAB / "tab_theil_provincia_anual.csv", index=False)

# ── Cut (b): rubro Shannon H per province x era ──────────────────────────────
def shannon(counts):
    p = counts / counts.sum()
    p = p[p > 0]
    return float(-(p * np.log(p)).sum())


k_rubros = adj["rubro_principal"].nunique()
rows = []
for (prov, era), g in adj.groupby(["provincia", "era"]):
    counts = g["rubro_principal"].value_counts()
    h = shannon(counts)
    rows.append({"provincia": prov, "era": era, "n_adj": len(g),
                 "n_rubros": len(counts), "shannon_h": h,
                 "shannon_norm": h / np.log(k_rubros),
                 "nea": prov in NEA})
tb = pd.DataFrame(rows)
tb.to_csv(TAB / "tab_shannon_rubros_provincia_era.csv", index=False)

# ── Cut (c): nested Theil by era ─────────────────────────────────────────────
rows = []
for era, g in ars.groupby("era"):
    firm = (g.groupby(["cuit", "provincia"], as_index=False)
            .agg(monto=("monto", "sum")))
    firm["region"] = np.where(firm["provincia"].isin(NEA), "NEA", "Resto")
    tot, bet_reg, _ = theil_decompose(firm, "monto", "region")
    _, bet_prov, wit_prov = theil_decompose(firm, "monto", "provincia")
    rows.append({"era": era, "theil_total": tot,
                 "between_region": bet_reg,
                 "between_provincia": bet_prov,
                 "between_prov_within_region": bet_prov - bet_reg,
                 "within_provincia": wit_prov})
tc = pd.DataFrame(rows).set_index("era").reindex(ERAS).reset_index()
tc.to_csv(TAB / "tab_theil_anidado_era.csv", index=False)

# ── Figures ──────────────────────────────────────────────────────────────────
fig, axes = plt.subplots(1, 2, figsize=(11, 4.2))
for ax, var, lab in zip(axes, ["monto", "n"],
                        ["Montos ARS", "Cantidad de adjudicaciones"]):
    d = ta[ta["variable"] == var]
    ax.plot(d["ejercicio"], d["theil_total"], "o-", label="Theil total", color="#333")
    ax.plot(d["ejercicio"], d["theil_between"], "s--", label="Between provincia", color="#d62728")
    ax2 = ax.twinx()
    ax2.bar(d["ejercicio"], 100 * d["share_between"], alpha=0.15, color="#1f77b4")
    ax2.set_ylabel("share between (%)", color="#1f77b4")
    for x in [2019.5, 2023.5]:
        ax.axvline(x, color="grey", lw=0.8, ls=":")
    ax.set_title(f"Theil T entre proveedores — {lab}")
    ax.set_xlabel("Ejercicio")
    ax.legend(fontsize=8)
fig.tight_layout()
fig.savefig(FIG / "fig_theil_temporal.png", dpi=200)
plt.close(fig)

fig, ax = plt.subplots(figsize=(8, 4.5))
for prov in sorted(NEA):
    d = tb[tb["provincia"] == prov].set_index("era").reindex(ERAS)
    ax.plot(ERAS, d["shannon_norm"], "o-", label=prov)
rest = (tb[~tb["nea"]].groupby("era")
        .apply(lambda g: np.average(g["shannon_norm"], weights=g["n_adj"]),
               include_groups=False).reindex(ERAS))
ax.plot(ERAS, rest.values, "k--", lw=2, label="Resto (prom. ponderado)")
ax.set_ylabel("Shannon H normalizado (rubros)")
ax.set_title("Diversidad de rubros de proveedores por era — NEA vs resto")
ax.legend(fontsize=8)
fig.tight_layout()
fig.savefig(FIG / "fig_shannon_nea.png", dpi=200)
plt.close(fig)

print("Cut (a) — share between-provincia por año (montos):")
print(ta[ta["variable"] == "monto"][["ejercicio", "theil_total", "share_between"]]
      .round(3).to_string(index=False))
print("\nCut (b) — Shannon normalizado, NEA por era:")
print(tb[tb["nea"]].pivot_table(index="provincia", columns="era",
                                values="shannon_norm").reindex(sorted(NEA))
      .round(3).to_string())
print("\nCut (c) — Theil anidado por era:")
print(tc.round(4).to_string(index=False))
print("OK -> tables/, figures/")

File: 03_theil_decomposition.R
# 03 — Theil decomposition of the national supplier market (canonical R port
# of 03_theil_decomposition.py; same outputs, same filenames).
# Run: Rscript 03_theil_decomposition.R   (from the project root)

source("theme_house.R", chdir = TRUE)

theil_t <- function(x) {
  x <- x[x > 0 & !is.na(x)]
  if (length(x) < 2) return(NA_real_)
  mu <- mean(x)
  mean((x / mu) * log(x / mu))
}

theil_decompose <- function(df, value, group) {
  d <- df[df[[value]] > 0 & !is.na(df[[value]]), ]
  if (nrow(d) < 2) return(c(total = NA, between = NA, within = NA))
  mu <- mean(d[[value]]); n <- nrow(d)
  total <- theil_t(d[[value]])
  parts <- d |>
    group_by(.data[[group]]) |>
    summarise(ng = n(), mug = mean(.data[[value]]),
              tg = theil_t(.data[[value]]), .groups = "drop")
  between <- sum((parts$ng / n) * (parts$mug / mu) * log(parts$mug / mu))
  within <- sum((parts$ng / n) * (parts$mug / mu) *
                  ifelse(is.na(parts$tg), 0, parts$tg))
  c(total = total, between = between, within = within)
}

shannon <- function(counts) {
  p <- counts / sum(counts)
  p <- p[p > 0]
  -sum(p * log(p))
}

adj <- read_parquet(file.path(DIR_PROC, "adjudicaciones_clean.parquet")) |>
  filter(es_nuevo, !is.na(provincia))
ars <- adj |> filter(moneda == "ARS")

# ── Cut (a): yearly Theil across suppliers, between/within province ─────────
ta <- bind_rows(lapply(sort(unique(ars$ejercicio)), function(yr) {
  firm <- ars |> filter(ejercicio == yr) |>
    group_by(cuit, provincia) |>
    summarise(monto = sum(monto, na.rm = TRUE), n = n(), .groups = "drop")
  bind_rows(lapply(c("monto", "n"), function(v) {
    r <- theil_decompose(firm, v, "provincia")
    tibble(ejercicio = yr, variable = v, theil_total = r["total"],
           theil_between = r["between"], theil_within = r["within"],
           share_between = r["between"] / r["total"], n_firmas = nrow(firm))
  }))
}))
write_tab(ta, "tab_theil_provincia_anual.csv")

# ── Cut (b): rubro Shannon H per province x era ──────────────────────────────
k_rubros <- n_distinct(adj$rubro_principal, na.rm = TRUE)
tb <- adj |>
  filter(!is.na(rubro_principal)) |>
  count(provincia, era, rubro_principal) |>
  group_by(provincia, era) |>
  summarise(n_adj = sum(n), n_rubros = n(), shannon_h = shannon(n),
            .groups = "drop") |>
  mutate(shannon_norm = shannon_h / log(k_rubros), nea = provincia %in% NEA)
write_tab(tb, "tab_shannon_rubros_provincia_era.csv")

# ── Cut (c): nested Theil by era ─────────────────────────────────────────────
tc <- bind_rows(lapply(ERAS, function(e) {
  firm <- ars |> filter(era == e) |>
    group_by(cuit, provincia) |>
    summarise(monto = sum(monto, na.rm = TRUE), .groups = "drop") |>
    mutate(region = ifelse(provincia %in% NEA, "NEA", "Resto"))
  r_reg <- theil_decompose(firm, "monto", "region")
  r_prov <- theil_decompose(firm, "monto", "provincia")
  tibble(era = e, theil_total = r_reg["total"],
         between_region = r_reg["between"],
         between_provincia = r_prov["between"],
         between_prov_within_region = r_prov["between"] - r_reg["between"],
         within_provincia = r_prov["within"])
}))
write_tab(tc, "tab_theil_anidado_era.csv")

# ── Figures ──────────────────────────────────────────────────────────────────
d_a <- ta |> mutate(variable = recode(variable, monto = "Montos ARS",
                                      n = "Cantidad de adjudicaciones"))
p1 <- ggplot(d_a, aes(ejercicio)) +
  geom_col(aes(y = share_between * max(theil_total, na.rm = TRUE)),
           fill = "#1f77b4", alpha = 0.15) +
  geom_line(aes(y = theil_total, colour = "Theil total")) +
  geom_point(aes(y = theil_total, colour = "Theil total")) +
  geom_line(aes(y = theil_between, colour = "Between provincia"),
            linetype = "dashed") +
  geom_point(aes(y = theil_between, colour = "Between provincia"), shape = 15) +
  geom_vline(xintercept = c(2019.5, 2023.5), linetype = "dotted",
             colour = "grey50", linewidth = 0.4) +
  scale_colour_manual(values = c("Theil total" = "#333333",
                                 "Between provincia" = "#d62728")) +
  facet_wrap(~variable, nrow = 1) +
  labs(x = "Ejercicio", y = "Theil T",
       title = "Theil T entre proveedores, between/within provincia") +
  theme_house()
save_fig(p1, "fig_theil_temporal.png", width = 11, height = 4.2)

rest_w <- tb |> filter(!nea) |> group_by(era) |>
  summarise(shannon_norm = weighted.mean(shannon_norm, n_adj), .groups = "drop") |>
  mutate(provincia = "Resto (prom. ponderado)")
d_b <- bind_rows(tb |> filter(nea) |> select(provincia, era, shannon_norm),
                 rest_w) |>
  mutate(era = factor(era, levels = ERAS))
p2 <- ggplot(d_b, aes(era, shannon_norm, group = provincia,
                      colour = provincia)) +
  geom_line() + geom_point() +
  scale_colour_manual(values = c(Chaco = "#d62728", Corrientes = "#ff7f0e",
                                 Formosa = "#9467bd", Misiones = "#8c564b",
                                 `Resto (prom. ponderado)` = "black")) +
  labs(x = NULL, y = "Shannon H normalizado (rubros)",
       title = "Diversidad de rubros de proveedores por era, NEA vs resto") +
  theme_house()
save_fig(p2, "fig_shannon_nea.png")

cat("\nCut (a) share between (montos):\n")
print(ta |> filter(variable == "monto") |>
        transmute(ejercicio, theil_total = round(theil_total, 3),
                  share_between = round(share_between, 3)), n = 20)
cat("\nCut (c):\n")
print(tc |> mutate(across(where(is.numeric), ~round(.x, 4))))
